### ISSUE TITLE
InteractiveChat - Unhandled exception occured in onPacketSending

### ISSUE THREAD

**Issue by TheCiROMG** [PaperMC/Paper#5220]
### What behaviour is expected:
I expected to see my message or a player's message in the game

### What behaviour is observed:
Basically nothing (in the game) on the console appears this error
![image](https://user-images.githubusercontent.com/61605643/108711761-970d3700-74f4-11eb-866c-e53fec31cf1a.png)

### Steps/models to reproduce:
Put the Interactive Chat (new plugin for my server), open the server, send a random message, SEE THE CONSOLE

### Plugin list:
![image](https://user-images.githubusercontent.com/61605643/108706597-958c4080-74ed-11eb-9d4c-f3c4460451d2.png)
With InteractiveChat

### Paper version:
![image](https://user-images.githubusercontent.com/61605643/108712037-f8cda100-74f4-11eb-8630-4550cd120f90.png)

### Anything else:
Unhandled exception occured in onPacketSending(PacketEvent) for InteractiveChat
Parameters:net.minecraft.server.v1_16_R3.PacketPlayOutChat@6bd113d9[

**Comment by Proximyst** [comment]
FOSS plugin: https://www.spigotmc.org/resources/interactivechat-supports-bungeecord-custom-keywords-items-inventories-in-chat-rgb-fonts.75870/

**Comment by kashike** [comment]
```
Caused by: java.lang.IllegalArgumentException: Cannot clone TextComponentImpl{.......} ( class net.kyori.adventure.text.TextComponentImpl): No cloner is suitable.
        at com.comphenix.protocol.reflect.cloning.AggregateCloner.clone(AggregateCloner.java:267) ~[?:?]
        at com.comphenix.protocol.reflect.cloning.NullableCloner.clone(NullableCloner.java:43) ~[?:?]
        at com.comphenix.protocol.reflect.cloning.FieldCloner.defaultTransform(FieldCloner.java:67) ~[?:?]
        at com.comphenix.protocol.events.PacketContainer$2$1.access$200(PacketContainer.java:1153) ~[?:?]
        at com.comphenix.protocol.events.PacketContainer$2$1$1.transformField(PacketContainer.java:1162) ~[?:?]
        at com.comphenix.protocol.reflect.ObjectWriter.copyToInternal(ObjectWriter.java:115) ~[?:?]
        at com.comphenix.protocol.reflect.ObjectWriter.copyTo(ObjectWriter.java:80) ~[?:?]
        at com.comphenix.protocol.reflect.cloning.FieldCloner.clone(FieldCloner.java:91) ~[?:?]
        at com.comphenix.protocol.reflect.cloning.AggregateCloner.clone(AggregateCloner.java:260) ~[?:?]
```
I think this is something @dmulloy2 needs to fix in ProtocolLib - I think this is a possible fix, but I have not tested it: https://paste.gg/p/kashike/2afe0bdc3d124d0b8b2592ec5126e12f

**Comment by dmulloy2** [comment]
Is Adventure a plugin or a new part of Paper? Not sure I love the idea of adding a random plugin as a dependency just to clone it

**Comment by kashike** [comment]
It's a part of Paper.

**Comment by Malfrador** [comment]
> Is Adventure a plugin or a new part of Paper? Not sure I love the idea of adding a random plugin as a dependency just to clone it

A library for component-based text formatting, and it's part of Paper since yesterday.

**Comment by LoneDev6** [comment]
Seems this lib introduced some problems with chat plugins, an example:
`java.lang.ClassCastException: class net.kyori.adventure.text.TextComponentImpl cannot be cast to class [Lnet.md_5.bungee.api.chat.BaseComponent; (net.kyori.adventure.text.TextComponentImpl and [Lnet.md_5.bungee.api.chat.BaseComponent; are in unnamed module of loader java.net.URLClassLoader @659499f1`

```
[17:51:10 ERROR]: [ItemsAdder] Unhandled exception occured in onPacketSending(PacketEvent) for ItemsAdder
java.lang.ClassCastException: class net.kyori.adventure.text.TextComponentImpl cannot be cast to class [Lnet.md_5.bungee.api.chat.BaseComponent; (net.kyori.adventure.text.TextComponentImpl and [Lnet.md_5.bungee.api.chat.BaseComponent; are in unnamed module of loader java.net.URLClassLoader @659499f1)
at dev.lone.itemsadder.g.e.onPacketSending(SourceFile:110) ~[?:?]
at com.comphenix.protocol.injector.SortedPacketListenerList.invokeSendingListener(SortedPacketListenerList.java:195) ~[?:?]
at com.comphenix.protocol.injector.SortedPacketListenerList.invokePacketSending(SortedPacketListenerList.java:149) ~[?:?]
at com.comphenix.protocol.injector.PacketFilterManager.handlePacket(PacketFilterManager.java:588) ~[?:?]
at com.comphenix.protocol.injector.PacketFilterManager.invokePacketSending(PacketFilterManager.java:564) ~[?:?]
at com.comphenix.protocol.injector.netty.ProtocolInjector.packetQueued(ProtocolInjector.java:336) ~[?:?]
at com.comphenix.protocol.injector.netty.ProtocolInjector.onPacketSending(ProtocolInjector.java:298) ~[?:?]
at com.comphenix.protocol.injector.netty.ChannelInjector.processSending(ChannelInjector.java:378) ~[?:?]
at com.comphenix.protocol.injector.netty.ChannelInjector.access$800(ChannelInjector.java:64) ~[?:?]
at com.comphenix.protocol.injector.netty.ChannelInjector$3.handleScheduled(ChannelInjector.java:343) ~[?:?]
at com.comphenix.protocol.injector.netty.ChannelInjector$3.onMessageScheduled(ChannelInjector.java:313) ~[?:?]
at com.comphenix.protocol.injector.netty.ChannelProxy$2.schedulingRunnable(ChannelProxy.java:127) ~[?:?]
at com.comphenix.protocol.injector.netty.EventLoopProxy.execute(EventLoopProxy.java:95) ~[?:?]
at net.minecraft.server.v1_16_R3.NetworkManager.b(NetworkManager.java:293) ~[paper-1.16.5.jar:git-Paper-478]
at net.minecraft.server.v1_16_R3.NetworkManager.dispatchPacket(NetworkManager.java:249) ~[paper-1.16.5.jar:git-Paper-478]
at net.minecraft.server.v1_16_R3.NetworkManager.sendPacket(NetworkManager.java:225) ~[paper-1.16.5.jar:git-Paper-478]
at net.minecraft.server.v1_16_R3.PlayerConnection.a(PlayerConnection.java:1738) ~[paper-1.16.5.jar:git-Paper-478]
at net.minecraft.server.v1_16_R3.PlayerConnection.sendPacket(PlayerConnection.java:1711) ~[paper-1.16.5.jar:git-Paper-478]
at org.bukkit.craftbukkit.v1_16_R3.entity.CraftPlayer.sendMessage(CraftPlayer.java:2130) ~[paper-1.16.5.jar:git-Paper-478]
at net.kyori.adventure.audience.Audience.sendMessage(Audience.java:274) ~[paper-1.16.5.jar:git-Paper-478]
at net.kyori.adventure.audience.Audience.sendMessage(Audience.java:195) ~[paper-1.16.5.jar:git-Paper-478]
at net.minecraft.server.v1_16_R3.PlayerConnection.chat(PlayerConnection.java:1999) ~[paper-1.16.5.jar:git-Paper-478]
at net.minecraft.server.v1_16_R3.PlayerConnection.c(PlayerConnection.java:1851) ~[paper-1.16.5.jar:git-Paper-478]
at net.minecraft.server.v1_16_R3.PlayerConnection.a(PlayerConnection.java:1789) ~[paper-1.16.5.jar:git-Paper-478]
at net.minecraft.server.v1_16_R3.PacketPlayInChat$1.run(PacketPlayInChat.java:41) ~[paper-1.16.5.jar:git-Paper-478]
at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515) ~[?:?]
at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
at java.lang.Thread.run(Thread.java:834) [?:?]
[17:51:10 ERROR]: Parameters:
net.minecraft.server.v1_16_R3.PacketPlayOutChat@2c8cb686[
a=<null>
components=<null>
b=SYSTEM
c=03e9d475-c2b0-4d8c-9223-db3741b260ed
] 

```

**Comment by electronicboy** [comment]
You'd need to provide some form of source or something, all we can see is a bad assumption on your side, not too much we can advise around closed source plugins

**Comment by Proximyst** [comment]
> at dev.lone.itemsadder.g.e.onPacketSending(SourceFile:110) ~[?:?]

That is literally caused by your own plugin doing stupid shit.

**Comment by LoneDev6** [comment]
> 
> 
> > at dev.lone.itemsadder.g.e.onPacketSending(SourceFile:110) ~[?:?]
> 
> That is literally caused by your own plugin doing stupid shit.

Thanks mate, very kind and adds a lot to the discussion



> 
> 
> You'd need to provide some form of source or something, all we can see is a bad assumption on your side, not too much we can advise around closed source plugins

You're right, I will provide the code in some hours, thanks for the patience.
Anyway the issue seems to be related to the fact that chat packets BaseComponent parameter does not inherit BaseComponent anymore and it's replaced by TextComponentImpl, this surely will affect other chat plugins that need to work with that object.

**Comment by MiniDigger** [comment]
It's generally that we tend to get frustrated by stuff like this. We do api changes and plugins that hack their way thru the internals without any safety checks break, and then suddenly ppl think we need to fix anything. 
We will however happily assist devs in fixing their implementation, which is obviously easier for open source plugins or libs like the ones mentioned in this issue, vs obfuscated proprietary paid stuff.
If ppl would actually use the api we would not have to deal with shit like this....

**Comment by LoneDev6** [comment]
> 
> 
> It's generally that we tend to get frustrated by stuff like this. We do api changes and plugins that hack their way thru the internals without any safety checks break, and then suddenly ppl think we need to fix anything.
> We will however happily assist devs in fixing their implementation, which is obviously easier for open source plugins or libs like the ones mentioned in this issue, vs obfuscated proprietary paid stuff.
> If ppl would actually use the api we would not have to deal with shit like this....

I appreciate your kindness and I totally see the point of that reaction.
I will provide the code as soon as I can, thanks again

**Comment by LoneDev6** [comment]
This code uses ProtocolLib + com.comphenix.protocol.wrappers

```java
new PacketAdapter(plugin, ListenerPriority.MONITOR, PacketType.Play.Server.CHAT)
                {
                    @Override
                    public void onPacketSending(PacketEvent e)
                    {
                        PacketContainer packet = e.getPacket();
                        if (packet.getChatComponents() == null)
                            return;

                        List<WrappedChatComponent> components = packet.getChatComponents().getValues();
                        //to fix Kingdoms plugin nonsense
                        if (components != null && components.size() != 0 && components.get(0) != null)
                        {
                            for (WrappedChatComponent component : components)
                            {
                                if (component == null)
                                    continue;
                                component.setJson(replaceFontImages(component.getJson()));
                                packet.getChatComponents().write(components.indexOf(component), component);
                            }
                        }
                        else
                        {
                            //here is the issue, I can't cast to BaseComponent anymore
                            BaseComponent[] baseComponents = (BaseComponent[]) e.getPacket().getModifier().read(1);
                            if (baseComponents == null || baseComponents.length == 0)
                                return;
                            if (baseComponents[0] == null)
                                return;
                            String json = ComponentSerializer.toString(baseComponents);
                            json = replaceFontImages(json);
                            baseComponents = ComponentSerializer.parse(json);
                            e.getPacket().getModifier().write(1, baseComponents);
                        }
                    }
                };
```

**Comment by jpenilla** [comment]
What is `replaceFontImages` doing? If it's replacing certain text content, a possible solution could be using instanceof to check if it's a `BaseComponent[]` (spigot) or a `Component` (paper), and on paper using the `Component#replaceText` method to create a modified component instead of working with the json directly.

**Comment by LoneDev6** [comment]
yes it's replacing text

holy crap, I didn't know about that method, I will check it for sure, thanks for the suggestion.
Is there any alternative to the `Component#replaceText` for Spigot? So that I can code two handlers (one for Spigot and one for Paper)

**Comment by jpenilla** [comment]
I am not familiar enough with the Bungeecord chat API to tell you whether or not it has a similar method, however I kind of doubt it if I'm being honest. I would suggest using your existing solution as a fallback on spigot.

**Comment by LoneDev6** [comment]
Here is my method if anyone needs that:
```java
if(e.getPacket().getModifier().read(1) instanceof TextComponent)
        {
            e.getPacket().getModifier().write(1, ((TextComponent) e.getPacket().getModifier().read(1)).replaceText(
                    TextReplacementConfig.builder()
                            .match(MY_PATTERN)
                            .replacement(builder -> builder.content("YOUR NEW TEXT HERE"))
                            .build()
            ));
        }
```

**Comment by kashike** [comment]
Should be fixed once https://github.com/dmulloy2/ProtocolLib/pull/1103 is merged.

**Comment by kashike** [comment]
This should be fixed with ProtocolLib 4.6.1-SNAPSHOT - please let us know if this is still an issue after updating.
